File: web/core/ASC.Web.Core/Utility/UrlShortener.cs
/*
 *
 * (c) Copyright Ascensio System Limited 2010-2020
 * 
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * http://www.apache.org/licenses/LICENSE-2.0
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
*/


using System;
using System.Configuration;
using System.Linq;
using System.Net;
using System.Security.Cryptography;
using System.Text;
using System.Web;

using ASC.FederatedLogin.LoginProviders;
using ASC.Security.Cryptography;
using ASC.Web.Studio.Utility;

namespace ASC.Web.Core.Utility
{
    public interface IUrlShortener
    {
        string GetShortenLink(string shareLink);
    }

    public static class UrlShortener
    {
        public static bool Enabled { get { return !(Instance is NullShortener); } }

        private static IUrlShortener _instance;
        public static IUrlShortener Instance
        {
            get
            {
                if (_instance == null)
                {
                    if (BitlyLoginProvider.Enabled)
                    {
                        _instance = new BitLyShortener();
                    }
                    else if (ConfigurationManagerExtension.AppSettings.AllKeys.Contains("web.url-shortener"))
                    {
                        _instance = new OnlyoShortener();
                    }
                    else
                    {
                        _instance = new NullShortener();
                    }
                }

                return _instance;
            }
            set
            {
                _instance = value;
            }
        }
    }

    public class BitLyShortener : IUrlShortener
    {
        public string GetShortenLink(string shareLink)
        {
            return BitlyLoginProvider.GetShortenLink(shareLink);
        }
    }

    public class OnlyoShortener : IUrlShortener
    {
        private readonly string url;
        private readonly string internalUrl;
        private readonly byte[] sKey;

        public OnlyoShortener()
        {
            url = ConfigurationManagerExtension.AppSettings["web.url-shortener"];
            internalUrl = ConfigurationManagerExtension.AppSettings["web.url-shortener.internal"];
            sKey = MachinePseudoKeys.GetMachineConstant();

            if (!url.EndsWith("/"))
                url += '/';
        }

        public string GetShortenLink(string shareLink)
        {
            using (var client = new WebClient { Encoding = Encoding.UTF8 })
            {
                client.Headers.Add("Authorization", CreateAuthToken());
                return CommonLinkUtility.GetFullAbsolutePath(url + client.DownloadString(new Uri(internalUrl + "?url=" + HttpUtility.UrlEncode(shareLink))));
            }
        }

        private string CreateAuthToken(string pkey = "urlShortener")
        {
            using (var hasher = new HMACSHA1(sKey))
            {
                var now = DateTime.UtcNow.ToString("yyyyMMddHHmmss");
                var hash = Convert.ToBase64String(hasher.ComputeHash(Encoding.UTF8.GetBytes(string.Join("\n", now, pkey))));
                return string.Format("ASC {0}:{1}:{2}", pkey, now, hash);
            }
        }
    }

    public class NullShortener : IUrlShortener
    {
        public string GetShortenLink(string shareLink)
        {
            return null;
        }
    }
}

File: module/ASC.UrlShortener/app/urlShortener.js
/*
 *
 * (c) Copyright Ascensio System Limited 2010-2020
 * 
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * http://www.apache.org/licenses/LICENSE-2.0
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
*/


/*
 * ShortURL (https://github.com/delight-im/ShortURL)
 * Copyright (c) delight.im (https://www.delight.im/)
 * Licensed under the MIT License (https://opensource.org/licenses/MIT)
 */

/**
 * ShortURL: Bijective conversion between natural numbers (IDs) and short strings
 *
 * ShortURL.encode() takes an ID and turns it into a short string
 * ShortURL.decode() takes a short string and turns it into an ID
 *
 * Features:
 * + large alphabet (51 chars) and thus very short resulting strings
 * + proof against offensive words (removed 'a', 'e', 'i', 'o' and 'u')
 * + unambiguous (removed 'I', 'l', '1', 'O' and '0')
 *
 * Example output:
 * 123456789 <=> pgK8p
 */

 //'23456789bcdfghjkmnpqrstvwxyzBCDFGHJKLMNPQRSTVWXYZ-_'

var _alphabet = "5XzpDt6wZRdsTrJkSY_cgPyxN4j-fnb9WKBF8vh3GH72QqmLVCM",
	_base = _alphabet.length,
	_initial = _base * _base;

module.exports = {

	encode: function(num) {
		num += _initial;
		var str = '';
		while (num > 0) {
			str = _alphabet.charAt(num % _base) + str;
			num = Math.floor(num / _base);
		}
		return str;
	},

	decode: function(str) {
		var num = 0;
		for (var i = 0; i < str.length; i++) {
            let index = _alphabet.indexOf(str.charAt(i));
            if (index < 0) return null;
			num = num * _base + index;
		}
		return num - _initial;
    }

};